File: heart-beat/heart-beat.cc
class Node {
public:
    Node(): ip(""), timestamp(0), next(nullptr) {}
    Node(string &i, int t): ip(i), timestamp(t), next(nullptr) {}
    
public:
    string ip;
    int timestamp;
    Node* next;
};

class HeartBeat {
public:
    HeartBeat() {
        // initialize your data structure here.
        root = new Node();
        key_pa.clear();
    }

    // @param slaves_ip_list a list of slaves'ip addresses
    // @param k an integer
    // @return void
    void initialize(vector<string>& slaves_ip_list, int k) {
        // Write your code here
        rtt = k * 2;
        Node *pa = root;
        for (auto &ip : slaves_ip_list) {
            Node *cur = new Node(ip, 0);
            key_pa[ip] = pa;
            pa->next = cur;
            pa = cur;
        }
        tail = pa;
    }

    // @param timestamp current timestamp in seconds
    // @param slave_ip the ip address of the slave server
    // @return nothing
    void ping(int timestamp, string& slave_ip) {
        // Write your code here
        if (key_pa.find(slave_ip) != key_pa.end()) {
            Node *pa = key_pa[slave_ip];
            Node *cur = pa->next;
            cur->timestamp = timestamp;
            if (cur == tail) return;
            pa->next = cur->next;
            key_pa[cur->next->ip] = pa;
            tail->next = cur;
            key_pa[slave_ip] = tail;
            tail = cur;
            tail->next = nullptr;
        }
    }

    // @param timestamp current timestamp in seconds
    // @return a list of slaves'ip addresses that died
    vector<string> getDiedSlaves(int timestamp) {
        // Write your code here
        vector<string> ans;
        Node *pa = root->next;
        while (pa != nullptr) {
            if (pa->timestamp <= timestamp - rtt) {
                ans.push_back(pa->ip);
                pa = pa->next;
            } else break;
        }
        return ans;
    }

private:
    int rtt;
    Node *root, *tail;
    map<string, Node *> key_pa;
};

File: heart-beat/heart-beat.py
class Node:
    def __init__(self, ip, ts):
        self.ip = ip
        self.ts = ts
        self.next = None


class HeartBeat:
    def __init__(self):
        # initialize your data structure here
        self.ip_pa = dict()
        self.head = Node("", 0)
        self.tail = None
        self.rtt = 0

    # @param {str[]} slaves_ip_list a list of slaves'ip addresses
    # @param {int} k an integer
    # @return nothing
    def initialize(self, slaves_ip_list, k):
        # Write your code here
        self.rtt = k + k
        pa = self.head
        cur = None
        for ip in slaves_ip_list:
            cur = Node(ip, 0)
            pa.next = cur
            self.ip_pa[ip] = pa
            pa = pa.next
        self.tail = cur

    # @param {int} timestamp current timestamp in seconds
    # @param {str} slave_ip the ip address of the slave server
    # @return nothing
    def ping(self, timestamp, slave_ip):
        # Write your code here
        if slave_ip in self.ip_pa:
            pa = self.ip_pa[slave_ip]
            cur = pa.next
            cur.ts = timestamp
            if cur == self.tail:
                return
            pa.next = cur.next
            self.ip_pa[cur.next.ip] = pa
            self.tail.next = cur
            self.ip_pa[slave_ip] = self.tail
            self.tail = cur
            self.tail.next = None

    # @param {int} timestamp current timestamp in seconds
    # @return {str[]} a list of slaves'ip addresses that died
    def getDiedSlaves(self, timestamp):
        # Write your code here
        ans = []
        pa = self.head.next
        while pa:
            if pa.ts + self.rtt > timestamp:
                break
            ans.append(pa.ip)
            pa = pa.next
        return ans

